build-support/bin/cache_comparison.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import argparse
+import json
+import shlex
+import shutil
+import subprocess
+import sys
+from time import time
+
+
+def create_parser() -> argparse.ArgumentParser:
+    parser = argparse.ArgumentParser(
+        description=(
+            "A (remote) cache comparison tool, which automates testing a range of Pants "
+            "builds (each in isolated cache namespaces) against a range of sources."
+        )
+    )
+
+    parser.add_argument(
+        "-a",
+        "--args",
+        default="check lint test ::",
+        help="The arguments to test each source commit with.",
+    )
+    parser.add_argument(
+        "-b",
+        "--build-diffspec",
+        help="The diffspec (e.g.: `main~10..main`) which selects the Pants builds to compare.",
+    )
+    parser.add_argument(
+        "--build-diffspec-step",
+        default=1,
+        help="The number of commits to step by within `--build-diffspec`.",
+    )
+    parser.add_argument(
+        "-s",
+        "--source-diffspec",
+        help=(
+            "The diffspec (e.g.: `main~10..main`) which selects the Pants-repo source commits "
+            "to run each Pants build against."
+        ),
+    )
+    parser.add_argument(
+        "--source-diffspec-step",
+        default=1,
+        help="The number of commits to step by within `--source-diffspec`.",
+    )
+    return parser
+
+
+def main() -> None:
+    args = create_parser().parse_args()
+    build_commits = commits_in_range(args.build_diffspec, args.build_diffspec_step)
+    source_commits = commits_in_range(args.source_diffspec, args.source_diffspec_step)
+    timings = timings_by_build(
+        shlex.split(args.args),
+        build_commits,
+        source_commits,
+    )
+    json.dump(timings, indent=2, fp=sys.stdout)
+
+
+Commit = str
+
+
+TimeInSeconds = float
+
+
+def commits_in_range(diffspec: str, step: int) -> list[Commit]:
+    all_commits = list(
+        subprocess.run(
+            ["git", "log", "--format=format:%h", diffspec, "--"],
```

Comment:
You could just `git rev-list` the diffspec. Maybe throw in `--reverse` instead of Pythoning that too.

Reply:
Thanks: this will almost certainly not pass on the first go, so will fix in a followup.